### PR TITLE
LibWeb: Add HTMLFormControlsCollection and RadioNodeList

### PR DESCRIPTION
**Pull request by jdahlin**
This contains the initial plumbing for supporting HTMLFormControlsCollection and RadioNodeList. HTMLFormsControlsCollection has an initial implementation but RadioNodeList is a stub.

### CODE REVIEW

File: Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+Variant<JS::Handle<RadioNodeList>, JS::Handle<DOM::Element>, Empty> HTMLFormControlsCollection::named_item(FlyString const& name)
+{
+    // 1. If name is the empty string, return null and stop the algorithm.
+    if (name.is_empty())
+        return {};
+
+    // 2. If, at the time the method is called, there is exactly one node in the collection
+    // that has either an id attribute or a name attribute equal to name, then return that node and stop the algorithm.
+    auto elements = root()->children();
+    if (elements->length() == 1) {
```
Comment (stelar7) [inline Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp:36]: Should this be 1 node total? or 1 node matching the id or name?

Comment (jdahlin) [reply Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp:36]: I read it as exactly one node that also matches id or name and the implementation is supposed to reflect that.

Comment (davidot) [reply Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp:36]: Testing this in firefox, it does seem like it means if there is just one element with that name/id. Not only if there is only one element

File: Userland/Libraries/LibWeb/HTML/RadioNodeList.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    return "";
+}
+
+JS::Value RadioNodeList::item_value(size_t index) const
+{
+    auto* node = item(index);
+    if (!node)
+        return JS::js_undefined();
+    return static_cast<DOM::Element const*>(node);
+}
```
Comment (davidot) [inline Userland/Libraries/LibWeb/HTML/RadioNodeList.cpp:45]: Why override this is it has the same implementation as the base class?

File: Userland/Libraries/LibWeb/HTML/RadioNodeList.cpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+WebIDL::ExceptionOr<void> RadioNodeList::set_value(String const& name)
+{
+    (void)name;
```
Comment (davidot) [inline Userland/Libraries/LibWeb/HTML/RadioNodeList.cpp:49]: Why not also already add the spec link (even though it is the same as for `value()`) and spec comments here?

I don't mind either way but doing one but not the other is not very consistent.

**Review by davidot — CHANGES_REQUESTED** [review]
Welcome to Serenity!

Great work to get started on this. I have a couple of small comments.

File: Userland/Libraries/LibWeb/HTML/RadioNodeList.cpp (hunk 2 above)
Comment (jdahlin) [reply Userland/Libraries/LibWeb/HTML/RadioNodeList.cpp]: Oversight, meant to implement more but I'll do that in another PR

File: Userland/Libraries/LibWeb/HTML/RadioNodeList.cpp (hunk 3 above)
Comment (jdahlin) [reply Userland/Libraries/LibWeb/HTML/RadioNodeList.cpp]: Good point, added.

File: Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+// https://html.spec.whatwg.org/multipage/common-dom-interfaces.html#dom-htmlformcontrolscollection-nameditem
+Variant<JS::Handle<RadioNodeList>, JS::Handle<DOM::Element>, Empty> HTMLFormControlsCollection::named_item(FlyString const& name)
+{
+    // 1. If name is the empty string, return null and stop the algorithm.
+    if (name.is_empty())
+        return {};
+
+    // 2. If, at the time the method is called, there is exactly one node in the collection
+    // that has either an id attribute or a name attribute equal to name, then return that node and stop the algorithm.
```
Comment (davidot) [inline Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp:34]: I still think the spec means something different here, but I guess this functions is still kind of TODO so perhaps it's fine

**Review by davidot — APPROVED** [review]
Code looks fine, just one comment about not matching how I read the spec there. But I think we can just use this as an initial step

File: Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp (hunk 4 above)
Comment (jdahlin) [reply Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp:34]: I've read it over a couple of times, how would you interpret it differently? AFAIU, root()->children() will give the current live collection. The spec mentions `The HTMLFormControlsCollection interface is used for collections of listed elements in form elements.`, so is_listed() has to be true for an element to be considered part of the collection.

Comment (davidot) [reply Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp:34]: I think that if there are multiple elements in the collection but just one has a name/id equal to `name` then it should return just that element. But I could also be misunderstand it, it is a bit ambiguous

Comment (jdahlin) [reply Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp:34]: I see, that is how I interpret the spec as well. Multiple elements is covered in step 4, which is marked as a TODO as it also involved RadioNodeList to contain these results. I plan to spend some time writing a few good test cases and try to implement the remaining parts if I can in a follow up PR.

Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp:34]: indent multi-line comments to the width of the numbering marker:

```
// 2. Foo
//    Bar
```

File: Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            if (!form_associated_element->is_listed()) {
+                auto& html_element = form_associated_element->form_associated_element_to_html_element();
+                if (html_element.get_attribute(AttributeNames::id) == name || html_element.get_attribute(AttributeNames::name) == name)
+                    return JS::make_handle(child);
+            }
+        }
+    }
+
+    // 3. Otherwise, if there are no nodes in the collection that have either an id attribute or a name
+    // attribute equal to name, then return null and stop the algorithm.
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp:49]: Likewise

File: Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            break;
+        }
+    }
+    if (!found) {
+        return {};
+    }
+
+    // 4. Otherwise, create a new RadioNodeList object representing a live view of the HTMLFormControlsCollection object,
+    // further filtered so that the only nodes in the RadioNodeList object are those that have either an id attribute or
+    // a name attribute equal to name. The nodes in the RadioNodeList object must be sorted in tree order.
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp:70]: Likewise

File: Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    TODO();
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.cpp:71]: Don't crash the engine on random unimplemented FIXMEs, it makes development extremely frustrating

File: Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.idl [hunk 8]
```diff
@@ -0,0 +1,8 @@
+#import <DOM/HTMLCollection.idl>
+#import <HTML/RadioNodeList.idl>
+
+[Exposed=Window]
+interface HTMLFormControlsCollection : HTMLCollection {
+  getter (RadioNodeList or Element)? namedItem(DOMString name);
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/HTMLFormControlsCollection.idl:6]: 4 spaces for indentation

File: Userland/Libraries/LibWeb/HTML/RadioNodeList.cpp [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    // 1. Let element be the first element in tree order represented by the RadioNodeList object
+    // that is an input element whose type attribute is in the Radio Button state and whose
+    // checkedness is true. Otherwise, let it be null.
+
+    // 2. If element is null, return the empty string.
+
+    // 3. If element is an element with no value attribute, return the string "on".
+
+    // 4. Otherwise, return the value of element's value attribute.
+    TODO();
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/RadioNodeList.cpp:35]: Likewise, remove

File: Userland/Libraries/LibWeb/HTML/RadioNodeList.cpp [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    // by the RadioNodeList object that is an input element whose type attribute is in the Radio Button state and
+    // whose value content attribute is either absent, or present and equal to the new value, if any.
+    // If no such element exists, then instead let element be null.
+
+    // Otherwise: let element be the first element in tree order represented by the RadioNodeList object that is
+    // an input element whose type attribute is in the Radio Button state and whose value content attribute is
+    // present and equal to the new value, if any. If no such element exists, then instead let element be null.
+
+    // 2. If element is not null, then set its checkedness to true.
+    TODO();
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/RadioNodeList.cpp:53]: Likewise, remove

File: Userland/Libraries/LibWeb/HTML/RadioNodeList.cpp [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+{
+}
+
+RadioNodeList::~RadioNodeList() = default;
+
+// https://html.spec.whatwg.org/multipage/common-dom-interfaces.html#dom-radionodelist-value
+String RadioNodeList::value() const
+{
+    // 1. Let element be the first element in tree order represented by the RadioNodeList object
+    // that is an input element whose type attribute is in the Radio Button state and whose
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/RadioNodeList.cpp:27]: Same here regarding comments, and below.

File: Userland/Libraries/LibWeb/HTML/RadioNodeList.idl [hunk 12]
```diff
@@ -0,0 +1,6 @@
+#import <DOM/NodeList.idl>
+
+[Exposed=Window]
+interface RadioNodeList : NodeList {
+  attribute DOMString value;
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/RadioNodeList.idl:5]: 4 spaces

**Review by linusg — CHANGES_REQUESTED** [review]
Please split this into two commits, having `and` in a commit message is more often than not a sign that you're not doing atomic commits.

**Pull request closed without merging**